### PR TITLE
Add better support for errors

### PR DESCRIPTION
**Pull request by danzimm**
This is needed for better error reporting. Currently when a user is already signed up and someone tries to sign up with the same email address I am only getting the following error:
```
Underlying(Error Domain=com.alamofire.error Code=-6003 "Response status code was unacceptable: 400" UserInfo={NSLocalizedFailureReason=Response status code was unacceptable: 400})`
```
And from this I can't figure out why the request failed.

I added more data to be passed through to other errors besides `Underlying` for the sake of ease of debugging (all the time the server team asks for the json that we get back to help debug an issue - with this change that will be easier to achieve). It's possible that in the future we will want an alternative way of passing the response and data down the chain since it's possible that the server team will want to see the raw headers/JSON even when we successfully parse the data (i.e. on a Success).

### CODE REVIEW

**Comment by JRHeaton** [comment]
I feel a bit iffy about these changes, as this is a repetitive interface change simply for debugging, where we already have a mechanism for such. When you call `CocoaBloc.Client.request()` you're given back an `Alamofire.Request`, on which you can attach any number of response handlers/serializers, including adding a handler for the full `Alamofire.Response<>` which will contain the `NSData` and `NSHTTPURLResponse` we seek to inspect.

**Comment by JRHeaton** [comment]
I see the usage friction with what I'm mentioning if we're using ReactiveCocoa. We should talk about the best way to improve it.

**Comment by danzimm** [comment]
For record keeping, we spoke offline and came to the conclusion of adding an optional `requestConfiguration: (Alamofire.Request -> ())` parameter to `request` that will allow developers to add response handlers to the request.

**Comment by JRHeaton** [comment]
We needn't actually put the `requestConfiguration` parameter in CocoaBloc – only ReactiveCocoaBloc. This function already returns the `Request`.

File: old/Extensions/Alamofire-Request+MantleSerialization.swift [hunk 1]
```diff
@@ -64,7 +80,23 @@ public extension Request {
...
+							let errorString = metadata["error"] as? String else {
+								return .Failure(Error.UnexpectedResponseType)
+						}
+						return .Failure(Error.API(errorString))
+					} else {
+						return .Failure(Error.UnexpectedResponseType)
+					}
+				} catch let error as NSError {
+					return .Failure(Error.JSONSerialization(error))
+				}
```
Comment (JRHeaton) [inline old/Extensions/Alamofire-Request+MantleSerialization.swift]: Any way we can unify some of this repetitive code between the two serializer functions? Maybe abstract it out into a private func?

**Comment by danzimm** [comment]
Good point, didn't realize there was a return value on those functions. Fixed now!

**Pull request merged**